include/numerics/petsc_vector.h

```diff
@@ -675,15 +674,17 @@ void PetscVector<T>::init (const numeric_index_type n,
 #ifdef LIBMESH_HAVE_MPI
       PetscInt petsc_n_local=cast_int<PetscInt>(n_local);
       libmesh_assert_less_equal (n_local, n);
-      ierr = VecCreateMPI (this->comm().get(), petsc_n_local, petsc_n,
-                           &_vec);
-      LIBMESH_CHKERR(ierr);
+      // Let us a create vector but not specify a vector type
+      // By default, mpivec will be created.
+      // Motivation: to allow users specify vector type at runtime to
```

Comment:
What happens to all the users who *don't* specify vector type at runtime?  I guess PETSc must be smart enough now to infer from the last two `VecSetSizes` parameters which backend it should switch to?  That would be awesome if so.  Is it a recent feature or something we need to check version numbers for?

Reply:
Hi @roystgnr, I'm one of the PETSc developers who is working with @fdkong on a project, so I've taken a look at this PR. If users don't specify any vector type, then PETSc will default to creating a `Vec` of type `VECSTANDARD`, which is essentially exactly what would happen before @fdkong's proposed changes (`VECSTANDARD` defaults to a plain-old `VECSEQ` on one MPI process or a `VECMPI` on multiple MPI processes). The type will be set in the `VecSetFromOptions()` call. We generally recommend using `VecCreate()` as in Fande's changes, rather than using `VecCreateSeq()` or `VecCreateMPI()`, actually, since the right thing should then happen automatically based on the size of the MPI communicator.

I should note, however, that even when using `VecCreateSeq()` and `VecCreateMPI()`, if the user is asking for a GPU type via the PETSc options data base (that is, by the command line, in most cases), PETSc will still convert the `Vec` to the specified GPU type when `VecSetFromOptions()` is called. (I just verified this in a debugger.) So, the use of these isn't the cause of whatever issues @fdkong is running into with getting GPU types working. It looks like libMesh does create some vectors from a `DM`. If some of those vectors need to be set to GPU types, then you'll need to specify something like `-dm_vec_type cuda` (using whatever appropriate prefix is appropriate for your DMs, if you've set something). I don't think it's the case with libMesh, but if there are vectors coming from something like `MatCreateVecs()`, then you'll need to make sure that the matrix is one of the GPU types, or, if it is a `MATSHELL`, you may need to add some logic to set the corresponding GPU type appropriately by calling `MatShellSetVecType()`.

Reply:
>I should note, however, that even when using VecCreateSeq() and VecCreateMPI(), if the user is asking for a GPU type via the PETSc options data base (that is, by the command line, in most cases),

I have another half PR that apply different prefixes for Vec and ghostedVec.

The actual reason was that `VecCreateGhost` does not do correct things on GPU